File: src/algorithm_java/leetcode/Problem543.java
package algorithm.leetcode;

public class Problem543 {

    public static int treeDepth(TreeNode root) {
        if (root == null)
            return 0;
        return Math.max(treeDepth(root.left), treeDepth(root.right)) + 1;
    }

    public int diameterOfBinaryTree(TreeNode root) {
        if (root == null)
            return 0;
        int a = treeDepth(root.left);
        int b = treeDepth(root.right);
        int c = diameterOfBinaryTree(root.left);
        int d = diameterOfBinaryTree(root.right);
        int res = Math.max(a + b, Math.max(c, d));
        return res;
    }

}

File: src/algorithm_cpp/problem543.cpp
class Solution {
private:
    int maxlen;
public:
    int search(TreeNode* node) {
	if (node == nullptr) return 0;
	int leftlen = search(node->left);
	int rightlen = search(node->right);
	if (leftlen + rightlen + 1 > maxlen) maxlen = leftlen + rightlen + 1;
	return max(leftlen, rightlen) + 1; 
    }

    int diameterOfBinaryTree(TreeNode* root) {
	if (root == nullptr) return 0;
	maxlen = 0;
	search(root);
	return maxlen - 1;
    }
};
